sourcegraph/cody
# Autocomplete: Use streaming to early-terminate Anthropic requests

This adds a backward-compatible support for using the new streaming backend added in https://github.com/sourcegraph/sourcegraph/pull/55967.

It currently only works in Node.js because the node-fetch polyfill we use returns native Node streams and not unified browser streams. However given that browser support is not official right now, I decided that for now (especially until we know this experiment is giving us wins) it's not worth to add a separate implementation.

This all works by the new `stream` bool option on the server. When set and the server has the changes in https://github.com/sourcegraph/sourcegraph/pull/55967, we return an SSE event stream. On the client, we can check the header values for wether we got a streaming response and use that.

Architecture wise this is not super nice yet and duplicates some code of post-processing to find out if the request is already finished. There are three econditions to detect wether a streaming response is "fullfilled" right now:

1. A single-line completion can be terminated if we have one non-empty full line of response
2. A multi-line completion can be terminated if the multi-line truncation logic (which currently uses indentation but might later use tree sitter) starts to truncate the answer
3. 2. A multi-line completion can be terminated if a line inside the new completion matches one that is already inside the document

## Test plan

Tested in combination with https://github.com/sourcegraph/sourcegraph/pull/55967 and sourcegraph.com (which is behind this change)

 - Run [sourcegraph/sourcegraph#55967](https://github.com/sourcegraph/sourcegraph/pull/55967) as a local SG server
 - Set up Cody VS Code to connect to localhost
 - Check out https://github.com/sourcegraph/cody/pull/723
 - Start a dev instance and see that completions will work and might be faster
   - I've tested this with our completion testing tool but since we only have a handful of examples and don't do a bazillion test runs, there's a log of variance in the early results. 
   - We can roll this out as an A/B test very easily (cc @taras-yemets)
